File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode curr = root;
        
        while(curr!=null) {
            //IF our curr's val is smaller we need to search in right
            if(curr.val<p.val && curr.val < q.val) {
                curr = curr.right;
            }
            //IF our curr's val is greater we need to search in left
            else if(curr.val > p.val && curr.val > q.val) {
                curr = curr.left;
            }
            //ALREADY IN RANGE
            else return curr;
        }
        
        return curr;
    }
}

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        TreeNode* curr = root;
        
        while(curr) {
            //IF our curr is smaller go to right subtree
            if(curr->val < p->val && curr->val < q->val) {
                curr = curr->right;
            }
            // If our  curr is greater go to left subtree
            else if (curr->val > p->val && curr->val > q->val) {
                curr = curr->left;
            }
            else return curr;   
        }
        return curr;
    }
};